blog: Navigating the Open Source Seas: My LFX Mentorship Experience with the CNCF Strimzi Community (#380)
strimzi/strimzi.github.io: _posts/2023-08-20-my-lfx-mentorship-experience-with-the-cncf-strimzi-community.md

The mentors' feedback and code reviews were instrumental in refining my work and aligning it with Strimzi’s high standards.

This MQTT-Kafka Bridge will enable one-way communication and mapping between MQTT and Kafka topics.
The transformed messages will simultaneously be delivered to the Kafka cluster by the Kafka producer.
This mapping makes it possible to integrate MQTT-based devices into an Apache Kafka system with interoperability, flexibility, and scalability.
It would enable us to take advantage of the benefits of both MQTT and Kafka in our data processing and IoT system scenarios.

This is what I worked on during each week.

![Mentorship tasks](/assets/images/posts/2023-08-20-mentorship-tasks-timeline.png)

The Bridge relies on a Mapper component to map the incoming MQTT messages to a valid Kafka Record, providing:
- `Interoperability`, e.g., MQTT QoS and Kafka Producer Acks level, MQTT wildcards topics;
- `Flexibility`, the user sets his topic mapping rules;
- And `scalability`, the bridge supports n:m topic mapping.

The user-defined `“topic mapping rules”` is a JSON array containing rules to define how the user wants to map each MQTT topic to a Kafka Topic.

> Working on this project also helped me shape my understanding and gain hands-on experience with integration testing. Apart from integration testing.

Reviewer: Not sure if this  sentence at the end was intended or not
```suggestion
Working on this project also helped me shape my understanding and gain hands-on experience with integration testing. 
```

I also enjoyed deploying the Bridge: packaging the Java application, building a container image, pushing it to a registry, and deploying the container on Kubernetes.  

## Mentorship and collaboration

No doubt, my mentors were the best part of my mentorship journey. 
I have them as my sensei because they played a crucial role in shaping my understanding of the technical aspects and the open source culture.
Their patience in answering my queries, willingness to guide me through challenges, and encouragement to explore different approaches were invaluable.
This collaborative environment extended beyond my immediate mentors, as the Strimzi community was always ready to provide insights and support whenever I reached out.
And, of course, I am glad I can count on my mentors and the Strimzi community even after the mentorship.

> Through the LFX Mentorship program, honed my technical skills, project management capabilities, and collaboration proficiency.

Reviewer: ```suggestion
Through the LFX Mentorship program, I honed my technical skills, project management capabilities, and collaboration proficiency.
```

The experience of working alongside experienced professionals taught me best practices, code organization, and the importance of maintaining clean and well-documented code.

## The weekly meetings

The weekly meetings were the best part of the mentorship program.
I had the opportunity to interact with the mentors, which was a great learning experience.
The meetings were very productive, and I got to learn a lot from them.
During the meetings, we discussed the progress made, the challenges faced, and the next steps to be taken.
The mentors were very supportive and provided valuable feedback and guidance.

## The result

The result was a bridge that allows you to produce messages to Kafka topics using MQTT.
You can find the code [here](https://github.com/strimzi/strimzi-mqtt-bridge).

The below image illustrates the working of the MQTT-Kafka Bridge.
It shows an Arduino device with three different sensors publishing data through the bridge and a Kafka consumer consuming these data. 

> <!-- Image here -->

Author: I am preparing the image I will use it. When ready, I will change it.

> It shows an Arduino device with three different sensors publishing data through the bridge and the Kafka Consumer consuming these data. 

Reviewer: ```suggestion
It shows an Arduino device with three different sensors publishing data through the bridge and a Kafka consumer consuming these data. 
```

# Conclusion

Participating in the LFX Mentorship program with the Strimzi community was a transformative experience that expanded my horizons and solidified my passion for open-source development.
The guidance, mentorship, and camaraderie from the mentors and the Strimzi community were invaluable assets that propelled my growth.
Reflecting on this journey, I am grateful for the opportunity to contribute to a remarkable project while surrounded by a community that embodies the true spirit of open-source collaboration.
This experience has enriched my technical knowledge and inspired me to continue my journey as a proactive contributor to the open-source ecosystem. 

# What is next?

During and after this fantastic journey, many good things happened:

> - I founded the first [Angolan open-source community](https://github.com/angolaosc) to educate, promote and help students and professionals get involved in Open-source:

Reviewer: ```suggestion
- I founded the first [Angolan open-source community](https://github.com/angolaosc) to educate, promote and help students and professionals get involved in open-source development:
```

I started the community in July, and by the end of my internship (late August), we were able to:

>     - introduce open-source to 200+ individuals;

Reviewer: `open-source` is adjective but is being used as a noun here

```suggestion
    - introduce the concept of open-source to 200+ individuals;
```

    - helped more than 20 individuals make their first contribution; 
    - grow our audience on social media by 700+ followers on LinkedIn;
    - and a discord community with 200+ members. 
- Also, I was invited to present what I did in the mentorship during the Strimzi Community meeting.

> - Not enough, I applied to be a core organizer for the [Cloud Native - Community Group Luanda](https://community.cncf.io/cloud-native-luanda) and got accepted.

Reviewer: ```suggestion
- Even further, I applied to be a core organizer for the [Cloud Native - Community Group Luanda](https://community.cncf.io/cloud-native-luanda) and got accepted.
```

I have embarked on a new journey as Open-source Program Manager at [Angola Open-source Community](https://github.com/angolaosc), where I am working on building a community of open-source enthusiasts in Angola.

I am very grateful for the opportunity to be part of the Strimzi community and the LFX Mentorship program.

> I am looking forward to continuing my journey as a proactive contributor to the open-source ecosystem and contributing to the Strimzi community.

Author: ```suggestion
I am looking forward to continuing my journey as a proactive contributor to the open-source ecosystem and contributing to the Strimzi community.

If you are looking for mentorship opportunities with CNCF, I wrote a compreensive blog on how to apply and be successful.
You can find the blog [here](https://antonio-pedro2019z.medium.com/exploring-mentorship-opportunities-with-cncf-your-path-to-growth-1df69ebe3124)
```

The Linux Foundation Mentorship is a 12-week online program to empower students to contribute to open-source projects under the [Cloud Native Computing Foundation](https://www.cncf.io/).

On the other hand, Strimzi is a [Cloud Native Computing Foundation](https://www.cncf.io/) sandbox project that focuses on providing operators for running Apache Kafka on Kubernetes.
With my interest piqued by this innovative endeavor, I decided to join the Strimzi community through the LFX Mentorship program.

The LFX mentorship program bridges budding developers and experienced mentors, providing an avenue to contribute meaningfully to open-source projects.

# From where everything started

Everything started in 2022, I was looking for things I could do during my summer holidays.
I investigated some mentorship programs such as Google Summer of Code, MLH Fellowship, Outreachy, and LFX Mentorship.

## The Application process

I have selected two projects for which my skill set fits the recommended skills.
Among them, the Strimzi project was there waiting for me. 

`“CNCF - Strimzi: Proof of Concept of an MQTT to Apache Kafka bridge for producing messages.”`

> This project got my attention among all the projects because I had practical experience with MQTT from my undergraduate program.

Reviewer 2: I would add the link to the project on the LFX website: https://mentorship.lfx.linuxfoundation.org/project/8d301adf-94d8-4e5d-821d-f904ed15c3f9

Author: I agree.

I learned the basic concepts of Apache Kafka from the Cloud Computing and Distributed System course at IIIT Delhi.

### The cover letter and the résumé

After understanding what I had to do and what the project was about, I had to prepare my cover letter and resume.
You need to submit these essential documents while applying to the LFX mentorship.
I ensured my cover letter and resume were well-aligned with the project by providing details about my skills and experience with the recommended skills.

But just after submitting my application, I noticed that one of the mentors for my project commented on the weak selection process.

> He was asking if they could ask for a proposal just like in Google Summer of Code.

Reviewer 2: ```suggestion
He was asking if the program should have a project design proposal as a prerequisite just like in Google Summer of Code.
A design proposal containing an overall description about the project, the challenges and related solutions.
A good place to start a discussion with mentors in order to understand how the mentee was approaching.
```

<!--Image here -->

This scared me :) But I was confident I could provide a good proposal because of my previous experience with the underlying technologies.

On the same day, I got my first interaction with my mentors, I would never think it wouldn’t be the last one.

<!--Image here -->

> They gave me two days to elaborate on the proposal, and to be honest, it was the first time I saw myself as a real engineer.

Reviewer 2: "They gave me two days" sounds a little bit harsh :-D as mentors were putting pressure on you.
It was just because of the "bad" LFX process which left us these two days for the proposal.
I would just remove this sentence.

Author: I see, fixed that.

The first task was understanding the real problem and why we would need to build such a bridge, so I researched that.
To design the solution, my first intuition was to relate my college project to the project.
Just to get a rough idea of how it is very similar to what I was building in college:

<!--Image here -->

After that, I started to think about the architecture of the solution, and I came up with this:

<!--Image here -->

To be even closer to a logical solution, the win-win was to study and understand how Strimzi HTTP-Kafka Bridge works and how it is deployed on Kubernetes. 
After reading the Strimzi HTTP-Kafka Bridge documentation, I was able to design a Bridge tackling all the issues. 
Of course, it was not perfect, and this is where discussions and critical thinking helped me.

> 

Reviewer 2: I would merge "From where everything started", "The Application process" and "The cover letter and the résumé" sections in just one. The first two are really small to deserve a title by themselves imho.
I would have just one maybe titled "The application process".

### The mentors feedback

After submitting my proposal, I got feedback from my mentors, and they were very positive about it.
They had a lot of comments asking about my approach, and discussing and resolving the conversions was, of course, part of my win-win.
At some point, I thought I wouldn’t be selected because my initial approach was mentioning unnecessary things like a full MQTT broker.

The feedback from my mentors allowed me to rethink and defend my approach, proving that I had a basic understanding and was keen to work on the project.

## The great day

After so many discussions on the right approach, feedback, etc., I was not sleeping waiting for my results. 
My first result was for the CNCF- Landing Page Improvement, in which I got rejected :(. 

> I started running unfaithfully and waiting for the Strimzi result, but it was too late, so I fell asleep hahaha.

Reviewer 2: ```suggestion
I started running unfaithfully and waiting for the Strimzi result, but it was too late, so I fell asleep.
```

I am not sure how the final "hahaha" would sound for a native english speaker. I would remove it.

### The work undertaken

The project I worked on was the MQTT-Kafka Bridge, which is a bridge that allows you to produce messages to Kafka topics using MQTT.
As I said before, I had to improve my solution, which could have been more manageable with many unnecessary components.
Still, with the guidance of mentors, I could break down the task into manageable components.
The mentors' feedback and code reviews were instrumental in refining my work and aligning it with Strimzi’s high standards.

This MQTT-Kafka Bridge will enable one-way communication and mapping between MQTT and Kafka topics.
The transformed messages will simultaneously be delivered to the Kafka cluster by the Kafka producer.
This mapping makes it possible to integrate MQTT-based devices into an Apache Kafka system with interoperability, flexibility, and scalability.
It would enable us to take advantage of the benefits of both MQTT and Kafka in our data processing and IoT system scenarios.

This is what I worked on during each week.

<!-- Image here -->

The Bridge relies on a Mapper component to map the incoming MQTT messages to a valid Kafka Record, providing:

> - `interoperability`, e.g., MQTT QoS and Kafka Producer Acks level, MQTT wildcards topics, 

Reviewer 2: This sentence ends with a comma, is there more missing?

Author: No, it is just a bad habit, I have replaced it with a `;`

- `flexibility`, the user sets his topic mapping rules

> - and `scalability`, the bridge supports n:m topic mapping.

Reviewer 2: I am not sure I see this as "scalability", it's just about n to m mapping.

Author: @ppatierno fixed! :-)

The user-defined `“topic mapping rules”` is a JSON array containing rules to define how the user wants to map each MQTT topic to a Kafka Topic.
Working on this project also helped me shape my understanding and gain hands-on experience with integration testing. Apart from integration testing.
I also enjoyed deploying the Bridge: packaging the Java application, building a container image, pushing it to a registry, and deploying the container on Kubernetes.  

### Mentorship and collaboration

No doubt, my mentors were the best part of my mentorship journey. 
I have them as my sensei because they played a crucial role in shaping my understanding of the technical aspects and the open source culture.

> Their patience in answering my queries, willingness to guide me through challenges, and encouragement to explore different approaches were invaluable. This collaborative environment extended beyond my immediate mentors, as the Strimzi community was always ready to provide insights and support whenever I reached out.

Reviewer 2: ```suggestion
Their patience in answering my queries, willingness to guide me through challenges, and encouragement to explore different approaches were invaluable.
This collaborative environment extended beyond my immediate mentors, as the Strimzi community was always ready to provide insights and support whenever I reached out.
```

New line ;-)

Author: I missed this one! :)

And, of course, I am glad I can count on my mentors and the Strimzi community even after the mentorship.

Through the LFX Mentorship program, honed my technical skills, project management capabilities, and collaboration proficiency.
The experience of working alongside experienced professionals taught me best practices, code organization, and the importance of maintaining clean and well-documented code.

### The weekly meetings

The weekly meetings were the best part of the mentorship program.
I had the opportunity to interact with the mentors, which was a great learning experience.

> The meetings were very productive, and I got to learn a lot from the mentors.

Reviewer 2: ```suggestion
The meetings were very productive, and I got to learn a lot from them.
```

During the meetings, we discussed the progress made, the challenges faced, and the next steps to be taken.
The mentors were very supportive and provided valuable feedback and guidance.

### The result

The result was a bridge that allows you to produce messages to Kafka topics using MQTT.

> You can find the code [here]().

Reviewer 2: missing link to the repo?

The guidance, mentorship, and camaraderie from the mentors and the Strimzi community were invaluable assets that propelled my growth.
Reflecting on this journey, I am grateful for the opportunity to contribute to a remarkable project while surrounded by a community that embodies the true spirit of open-source collaboration.
This experience has enriched my technical knowledge and inspired me to continue my journey as a proactive contributor to the open-source ecosystem. 

# What is next?

During and after this fantastic journey, many good things happened:
- I founded the first [Angolan open-source community](https://github.com/angolaosc) to educate, promote and help students and professionals get involved in Open-source:
I started the community in July, and by the end of my internship (late August), we were able to:
 - introduce open-source to 200+ individuals;
 - helped more than 20 individuals make their first contribution; 
 - grow our audience on social media by 700+ followers on LinkedIn;
 - and a discord community with 200+ members. 
- Not only that, but I was invited to present what I did in the mentorship during the Strimzi Community meeting. 

> I have embarked on a new journey as Open-source Program Manager at [Angola Open-source Community](), where I am working on building a community of open-source enthusiasts in Angola.

Reviewer 2: missed link?

Author: I am going to add the links.

I started the community in July, and by the end of my internship (late August), we were able to:
    - introduce the concept of open-source to 200+ individuals;
    - helped more than 20 individuals make their first contribution; 
    - grow our audience on social media by 700+ followers on LinkedIn;
    - and a discord community with 200+ members. 
- Also, I was invited to present what I did in the mentorship during the Strimzi Community meeting.
- Even further, I applied to be a core organizer for the [Cloud Native - Community Group Luanda](https://community.cncf.io/cloud-native-luanda) and got accepted.

I have embarked on a new journey as Open-source Program Manager at [Angola Open-source Community](https://github.com/angolaosc), where I am working on building a community of open-source enthusiasts in Angola.

I am very grateful for the opportunity to be part of the Strimzi community and the LFX Mentorship program.
I am looking forward to continuing my journey as a proactive contributor to the open-source ecosystem and contributing to the Strimzi community.

> If you are looking for mentorship opportunities with CNCF, I wrote a compreensive blog on how to apply and be successful.

Author: ```suggestion
If you are looking for mentorship opportunities with CNCF, I wrote a comprehensive guide on how to apply and be successful.
```

![Mentor's email asking for a proposal](/assets/images/posts/2023-08-20-mentor-email.png)

The first task was understanding the real problem and why we would need to build such a bridge, so I researched that.
To design the solution, my first intuition was to relate my college project to the project.
Just to get a rough idea of how similar it was to what I was building in college:

![Architecture of a system using MQTT](/assets/images/posts/2023-08-20-design-of-a-system-using-mqtt.png)

After that, I started to think about the architecture of the solution, and I came up with this:

![Strimzi MQTT Bridge](/assets/images/posts/2023-08-20-mqtt-bridge-high-architecture.png)

To be even closer to a logical solution, the win-win was to study and understand how Strimzi HTTP-Kafka Bridge works and how it is deployed on Kubernetes. 
After reading the Strimzi HTTP-Kafka Bridge documentation, I was able to design a Bridge tackling all the issues. 
Of course, it was not perfect, and this is where discussing my solution with the mentors helped me.

#### Mentor feedback

After submitting my proposal, I got feedback from my mentors, which was overwhelmingly positive.

> They had lots of comments about my approach,. Discussing and resolving their concerns was, of course, part of my win-win.

Author: @PaulRMellor I was talking about this. 
```suggestion
They had lots of comments about my approach. Discussing and resolving their concerns was, of course, part of my win-win.
```

At some point, I thought I wouldn’t be selected because my initial approach was mentioning unnecessary things like a full MQTT broker.

The feedback from my mentors allowed me to rethink and defend my approach, proving that I had a basic understanding and was keen to work on the project.

## The great day

After so many discussions on the right approach, feedback, etc., I was not sleeping waiting for my results. 
My first result was for the CNCF- Landing Page Improvement, in which I got rejected :(. 

I started running unfaithfully and waiting for the Strimzi result, but it was too late, so I fell asleep.

> Waking up too early, I opened the email, and the great news was there

Author: ```suggestion
Waking up too early, I opened the email, and the great news was there:
```

> 

Author: ```suggestion

![Acceptance email(/assets/images/posts/2023-08-20-acceptance-email.png)

```

You can find more information about the project [here](https://mentorship.lfx.linuxfoundation.org/project/8d301adf-94d8-4e5d-821d-f904ed15c3f9).

After understanding what I had to do and what the project was about, I had to prepare my cover letter and resume.
You need to submit these essential documents while applying to the LFX mentorship.
I ensured my cover letter and resume were well-aligned with the project by providing details about my skills and experience with the recommended skills.

But just after submitting my application, I noticed that one of the mentors for my project commented on the weak selection process.
He was asking if the program should have a project design proposal as a prerequisite just like in Google Summer of Code.
A design proposal containing an overall description about the project, the challenges and related solutions.
A good place to start a discussion with mentors in order to understand how the mentee was approaching.

![Mentor's GitHub comment](/assets/images/posts/2023-08-20-github-comment.png)

This scared me :) But I was confident I could provide a good proposal because of my previous experience with the underlying technologies.

> On the same day, I got my first interaction with my mentors, I would never think it wouldn’t be the last one.

Author: Yes, you are correct, this is what I meant.

![Mentor's email asking for a proposal](/assets/images/posts/2023-08-20-mentor-email.png)

The first task was understanding the real problem and why we would need to build such a bridge, so I researched that.
To design the solution, my first intuition was to relate my college project to the project.
Just to get a rough idea of how it is very similar to what I was building in college:

![Architecture of a system using MQTT](/assets/images/posts/2023-08-20-design-of-a-system-using-mqtt.png)

After that, I started to think about the architecture of the solution, and I came up with this:

![Strimzi MQTT Bridge](/assets/images/posts/2023-08-20-mqtt-bridge-high-architecture.png)

To be even closer to a logical solution, the win-win was to study and understand how Strimzi HTTP-Kafka Bridge works and how it is deployed on Kubernetes. 
After reading the Strimzi HTTP-Kafka Bridge documentation, I was able to design a Bridge tackling all the issues. 
Of course, it was not perfect, and this is where discussing about my solution with the mentors helped me.

#### The mentors feedback

After submitting my proposal, I got feedback from my mentors, and they were very positive about it.

> They had a lot of comments asking about my approach, and discussing and resolving the conversions was, of course, part of my win-win.

Author: ```suggestion
They had a lot of comments asking about my approach and discussing and resolving the conversions was, of course, part of my win-win.
```

> At some point, I thought I wouldn’t be selected because my initial approach was mentioning unnecessary things like a full MQTT broker.

Author: The only issue here is that you did not delete the `,` before adding the  `.`. But I got it.

---
layout: post
title: "Navigating the Open Source Seas: My LFX Mentorship Experience with the CNCF Strimzi Community"
date: 2023-08-20
author: antonio_pedro
---

In the world of open-source software, collaboration, and mentorship play a pivotal role in driving innovation and fostering growth. 
As an aspiring developer looking to dive into the intricate world of cloud-native technologies, I embarked on a remarkable journey by participating in the [LFX Mentorship](https://lfx.linuxfoundation.org/tools/mentorship) program with the [Strimzi](https://strimzi.io) community.
This program is happens thrice a year and is organized by [The Linux Foundation](https://www.linuxfoundation.org/).
This blog post aims to shed light on my enriching experience, highlighting the valuable insights gained, the impactful work undertaken, and the immeasurable support from the mentors and the Strimzi community.

<!--more-->

> # LFX Mentorship and the Cloud Native Computing Foundation Strimzi Community

Reviewer 3: The way I normally format the blog post headers is that these should be _level 3_ headers. It follows the logic that the page title (e.g. Strimzi) is level 1, the blog post title is level 2 heading and in your blog post you can use level 3 and level 4 for example. In other words, replace all the headers with `# ...` as `### ...` and the headers you have as `## ...` replace with `#### ...`.

When you check the preview of the blog post of how it renders on the website, it will get a much cleaner look with better spacing etc.: https://deploy-preview-380--strimzi.netlify.app/blog/2023/08/20/my-lfx-mentorship-experience-with-the-cncf-strimzi-community/

Author: Hi @scholzj, thanks for your feedback. 
I understood what I need to do, I will make the required changes.

> # LFX Mentorship and the Cloud Native Computing Foundation Strimzi Community

Author: ```suggestion
### LFX Mentorship and the Cloud Native Computing Foundation Strimzi Community
```

The Linux Foundation Mentorship is a 12-week online program to empower students to contribute to open-source projects under the [Cloud Native Computing Foundation](https://www.cncf.io/).

On the other hand, Strimzi is a [Cloud Native Computing Foundation](https://www.cncf.io/) sandbox project that focuses on providing operators for running Apache Kafka on Kubernetes.
With my interest piqued by this innovative endeavor, I decided to join the Strimzi community through the LFX Mentorship program.

The LFX mentorship program bridges budding developers and experienced mentors, providing an avenue to contribute meaningfully to open-source projects.

> # The application process

Author: ```suggestion
### The application process
```

On the same day, I got my first interaction with my mentors, I would never think it wouldn’t be the last one.

![Mentor's email asking for a proposal](/assets/images/posts/2023-08-20-mentor-email.png)

The first task was understanding the real problem and why we would need to build such a bridge, so I researched that.
To design the solution, my first intuition was to relate my college project to the project.
Just to get a rough idea of how it is very similar to what I was building in college:

![Architecture of a system using MQTT](/assets/images/posts/2023-08-20-design-of-a-system-using-mqtt.png)

After that, I started to think about the architecture of the solution, and I came up with this:

![Strimzi MQTT Bridge](/assets/images/posts/2023-08-20-mqtt-bridge-high-architecture.png)

To be even closer to a logical solution, the win-win was to study and understand how Strimzi HTTP-Kafka Bridge works and how it is deployed on Kubernetes. 
After reading the Strimzi HTTP-Kafka Bridge documentation, I was able to design a Bridge tackling all the issues. 
Of course, it was not perfect, and this is where discussing about my solution with the mentors helped me.

> ## The mentors feedback

Author: ```suggestion
#### The mentors feedback
```

After submitting my proposal, I got feedback from my mentors, and they were very positive about it.
They had a lot of comments asking about my approach, and discussing and resolving the conversions was, of course, part of my win-win.
At some point, I thought I wouldn’t be selected because my initial approach was mentioning unnecessary things like a full MQTT broker.

The feedback from my mentors allowed me to rethink and defend my approach, proving that I had a basic understanding and was keen to work on the project.

> ## The great day

Author: ```suggestion
#### The great day
```

After so many discussions on the right approach, feedback, etc., I was not sleeping waiting for my results. 
My first result was for the CNCF- Landing Page Improvement, in which I got rejected :(. 

I started running unfaithfully and waiting for the Strimzi result, but it was too late, so I fell asleep.
Waking up too early, I opened the email, and the great news was there:

> ![Acceptance email(/assets/images/posts/2023-08-20-acceptance-email.png)

Author: ```suggestion
![Acceptance email](/assets/images/posts/2023-08-20-acceptance-email.png)
```

> # The Mentorship journey

Author: ```suggestion
### The Mentorship journey
```

Upon joining the Strimzi community, I was warmly welcomed by mentors who were experts in their fields and exceptionally supportive and approachable.
With their sense of humor, they gave me comprehensive guidance on understanding the Strimzi project governance, setting up a repository for the project I worked on, etc.

> A big shout-out to [kyguy](https://github.com/kyguy) and [pppatierno](https://github.com/ppatierno) for being my mentors and guiding me throughout the mentorship journey.

Reviewer 2: Too many "p" in my GitHub handle :-D ... the link is fine, the text should be "pp" and not "ppp" :-P

Author: Sorry :)

This initial interaction set the tone for a productive and collaborative mentorship journey.

> ## The work undertaken

Author: ```suggestion
#### The work undertaken
```

> The project I worked on was the MQTT-Kafka Bridge, which is a bridge that allows you to produce messages to Kafka topics using MQTT.

Reviewer 2: I would end with "... but using the MQTT protocol"

This mapping makes it possible to integrate MQTT-based devices into an Apache Kafka system with interoperability, flexibility, and scalability.
It would enable us to take advantage of the benefits of both MQTT and Kafka in our data processing and IoT system scenarios.

This is what I worked on during each week.

![Mentorship tasks](/assets/images/posts/2023-08-20-mentorship-tasks-timeline.png)

The Bridge relies on a Mapper component to map the incoming MQTT messages to a valid Kafka Record, providing:
- `Interoperability`, e.g., MQTT QoS and Kafka Producer Acks level, MQTT wildcards topics;
- `Flexibility`, the user sets his topic mapping rules;
- And `scalability`, the bridge supports n:m topic mapping.

The user-defined `“topic mapping rules”` is a JSON array containing rules to define how the user wants to map each MQTT topic to a Kafka Topic.
Working on this project also helped me shape my understanding and gain hands-on experience with integration testing. 
I also enjoyed deploying the Bridge: packaging the Java application, building a container image, pushing it to a registry, and deploying the container on Kubernetes.  

> ## Mentorship and collaboration

Author: ```suggestion
#### Mentorship and collaboration
```

No doubt, my mentors were the best part of my mentorship journey. 
I have them as my sensei because they played a crucial role in shaping my understanding of the technical aspects and the open source culture.
Their patience in answering my queries, willingness to guide me through challenges, and encouragement to explore different approaches were invaluable.
This collaborative environment extended beyond my immediate mentors, as the Strimzi community was always ready to provide insights and support whenever I reached out.
And, of course, I am glad I can count on my mentors and the Strimzi community even after the mentorship.

Through the LFX Mentorship program, I honed my technical skills, project management capabilities, and collaboration proficiency.
The experience of working alongside experienced professionals taught me best practices, code organization, and the importance of maintaining clean and well-documented code.

> ## The weekly meetings

Author: ```suggestion
#### The weekly meetings
```

The weekly meetings were the best part of the mentorship program.
I had the opportunity to interact with the mentors, which was a great learning experience.
The meetings were very productive, and I got to learn a lot from them.
During the meetings, we discussed the progress made, the challenges faced, and the next steps to be taken.
The mentors were very supportive and provided valuable feedback and guidance.

> ## The result

Author: ```suggestion
#### The result
```

> The result was a bridge that allows you to produce messages to Kafka topics using MQTT.

Reviewer 2: "... using MQTT protocol"

You can find the code [here](https://github.com/strimzi/strimzi-mqtt-bridge).

The below video illustrates the working of the MQTT-Kafka Bridge.
It shows an Arduino device with three different sensors publishing data through the bridge and a Kafka consumer consuming these data. 

<iframe width="560" height="315" src="https://www.youtube.com/embed/VdNLLCVxbC8" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

> # Conclusion

Author: ```suggestion
### Conclusion
```

Participating in the LFX Mentorship program with the Strimzi community was a transformative experience that expanded my horizons and solidified my passion for open-source development.
The guidance, mentorship, and camaraderie from the mentors and the Strimzi community were invaluable assets that propelled my growth.
Reflecting on this journey, I am grateful for the opportunity to contribute to a remarkable project while surrounded by a community that embodies the true spirit of open-source collaboration.
This experience has enriched my technical knowledge and inspired me to continue my journey as a proactive contributor to the open-source ecosystem. 

> # What is next?

Author: ```suggestion
### What is next?
```
